### PR TITLE
Insert fronts banner ads on fronts

### PR DESCRIPTION
**Pull request by domlander**
## What does this change?

_Note: The following functionality is hidden behind a feature switch_

- Renders fronts-banner ads on fronts and tagged fronts pages instead of merchandising-high slots and MPUs.
- For high-traffic fronts that were used in the AB tests, ads are inserted above the collections as specified in this [previously defined file](https://github.com/guardian/dotcom-rendering/blob/main/dotcom-rendering/src/lib/frontsBannerAbTestAdPositions.ts).
- For fronts that were not used in the AB test, ads are inserted above every third collection.
- Creates a file `frontsBannerExclusions.ts` to keep track of collections we don't want to insert ads above. This will later be done by configuration in the fronts config, but for now we need this file to keep track.
- On tagged fronts for ad-free users, don't show whitespace where the ad would go. The other containers now stretch to fill the empty space. Tagged fronts are not yet rendered by DCR, so these changes will not yet be reflected on prod.
- The two files `TagFrontFastMpu` and `TagFrontSlowMpu` will be refactored later, as we will no longer show MPUs inside fronts containers. However, whilst we have the `frontsBannerAds` feature switch, these will need to retain their existing functionality of being able to display MPU ads, in case the feature switch is turned off.

## Why?

- Following successful [AB](https://github.com/guardian/dotcom-rendering/pull/8481) [tests](https://github.com/guardian/dotcom-rendering/pull/8703) and evaluating the effect on potential revenue, page performance and user experience metrics, we have decided to implement this change to ad display strategy on fronts pages.

### New type of ads on fronts

| Before      | After      |
| ----------- | ---------- |
| ![before][] | ![after][] |

[before]: https://github.com/guardian/dotcom-rendering/assets/9574885/62f7df18-6b03-44fb-9a8c-86424894a8dc
[after]: https://github.com/guardian/dotcom-rendering/assets/9574885/6ab333b9-2caf-469a-8b74-414a84a45f8e

### High traffic fronts (small amount of fronts that are responsible for majority of impressions)

| Before      | After      |
| ----------- | ---------- |
| ![before3][] | ![after3][] |

[before3]: https://github.com/guardian/dotcom-rendering/assets/9574885/1021f7c5-5aeb-4086-b60e-0add6297c394
[after3]: https://github.com/guardian/dotcom-rendering/assets/9574885/c77eb077-f657-45df-8a77-32f5d9d9f2e9

### Standard fronts

| Before      | After      |
| ----------- | ---------- |
| ![before4][] | ![after4][] |

[before4]: https://github.com/guardian/dotcom-rendering/assets/9574885/64aea22f-1b7c-4ed3-b4b2-38d217f674ec
[after4]: https://github.com/guardian/dotcom-rendering/assets/9574885/3499e2c8-4a11-4811-aff4-91f121e835a7

### Tagged fronts

| Before      | After      |
| ----------- | ---------- |
| ![before2][] | ![after2][] |

[before2]: https://github.com/guardian/dotcom-rendering/assets/9574885/49b9d8e2-254f-46d2-8a6e-e56ac1bcc5d9
[after2]: https://github.com/guardian/dotcom-rendering/assets/9574885/ec3d6926-67db-4221-b732-38cec80a9003

### CODE REVIEW

File: dotcom-rendering/src/lib/commercial-constants.ts [hunk 1]
```diff
@@ -0,0 +1,4 @@
+/**
+ * The maximum number of fronts-banner ads that can be inserted on any front.
+ */
+export const MAX_FRONTS_BANNER_ADS = 6;
```
Comment (domlander) [inline dotcom-rendering/src/lib/commercial-constants.ts:4]: Note that this applies to ads inserted via algorithm and is not applicable to pages that have a custom collection specification defined in `frontsBannerAbTestAdPositions.ts`

File: dotcom-rendering/src/layouts/lib/decideAdSlots.tsx [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+			);
+		}
+	}
+
+	// Insert an ad after every third collection. Warning: may skip an ad if a collection isn't rendered.
+	// e.g. if the 15th collection doesn't render, an ad is shown above the 12th and the 18th, skipping the 15th.
+	else if (
+		showBannerAds &&
+		numBannerAdsInserted.current < MAX_FRONTS_BANNER_ADS &&
+		index % 3 === 2 && // Insert above the 3rd, 6th, ..., container
```
Comment (Jakeii) [inline dotcom-rendering/src/layouts/lib/decideAdSlots.tsx:122]: This
```suggestion
		(index  + 1) % 3 === 0 && // Insert above the 3rd, 6th, ..., container
```
would be clearer in my head, but not a blocker

File: dotcom-rendering/src/layouts/lib/decideAdSlots.tsx [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+};
+
+/**
+ * Renders a fronts-banner ad when in the fronts banner AB test.
+ * Only applies to network fronts on desktop screens and wider.
+ */
+export const decideFrontsBannerAdSlot = (
+	renderAds: boolean,
+	hasPageSkin: boolean,
+	numBannerAdsInserted: React.MutableRefObject<number>,
```
Comment (Jakeii) [inline dotcom-rendering/src/layouts/lib/decideAdSlots.tsx:62]: Do we need this mutable ref? I think a plain variable in the scope of this new module would do the job?

**Review by Jakeii — APPROVED** [review]
Nicely done Dom! 🎉 Just a couple non-blocking comments

File: dotcom-rendering/src/layouts/lib/decideAdSlots.tsx (hunk 3 above)
Comment (domlander) [reply dotcom-rendering/src/layouts/lib/decideAdSlots.tsx:62]: Yeah I think you're right. Let's test this afterwards but I see no reason why we need a ref over a variable here.

File: dotcom-rendering/src/lib/commercial-constants.ts (hunk 1 above)
Comment (arelra) [reply dotcom-rendering/src/lib/commercial-constants.ts:4]: Minor - it looks like this is only used by one module so it seems a little unnecessary to be extracted to its own file.

**Review by emma-imber — APPROVED** [review]
Amazing stuff! 🎉

File: dotcom-rendering/src/layouts/lib/decideAdSlots.tsx (hunk 3 above)
Comment (arelra) [reply dotcom-rendering/src/layouts/lib/decideAdSlots.tsx:62]: [useRef](https://react.dev/reference/react/useRef) should be equivalent

Comment (arelra) [reply dotcom-rendering/src/layouts/lib/decideAdSlots.tsx:62]: refs don't force a re-render worth checking that you are not relying on it to update state used by the render.

https://react.dev/reference/react/useRef

**Review by arelra — APPROVED** [review]
Only minor non-blocking comments. Worth double checking your usage of refs.

Great work Dom! 🎉

File: dotcom-rendering/src/lib/commercial-constants.ts (hunk 1 above)
Comment (domlander) [reply dotcom-rendering/src/lib/commercial-constants.ts:4]: Yep, good spot

File: dotcom-rendering/src/components/TagFrontFastMpu.tsx [hunk 4]
```diff
@@ -10,57 +10,67 @@ import type { Loading } from './CardPicture';
...
 }: {
 	trails: Tuple<DCRFrontCard, 2>;
+	renderAds: boolean;
 	adIndex: number;
 	imageLoading: Loading;
 }) => {
 	return (
 		<UL direction="row">
-			<LI percentage="33.333%" padSides={true}>
+			<LI percentage={renderAds ? '33.333%' : '50%'} padSides={true}>
```
[the remark this answers is not in the document]
Comment (domlander) [reply dotcom-rendering/src/components/TagFrontFastMpu.tsx]: Decided not to make this change to tagged fronts as these haven't moved to DCR yet and require a design change. I didn't want to delay the release.

**Pull request merged**